File: TetJava/Blocks/BlockABC.java
package TetJava.Blocks;

import TetJava.Board;
import TetJava.KeyListener.Keylistener;
/**
 * This class will be extended by multiple block classes and will act as a framework for all of the block types to run.
 * This does not have a constructor, as it is an abstract class.
 */
public class BlockABC {
    /** This variable contains a coordinate value for the block relative to the board */
    protected int xpos, ypos;

    /** This will contain the location of the parts reletive to the whole body of the block */
    protected int[][] partCoords;

    /** This will contain all possible coordinate values for the variable orientations of the block */
    protected int[][][] orientations;

    /** This variable will store the last key that was pressed on the keyboard */
    protected int keyPressed;

    /** This will contain the farthest right x coordinate reletive to the block */
    protected int farthestFromBlockx = 0;

    /** This will contain the farthest right y coordinate reletive to the block */
    protected int farthestFromBlocky = 0;

    /** This will hold the current board instance */
    protected Board board;

    public BlockABC(Board board) {
        this.board = board;
    }

    /** Returns the last key that was pressed */
    private void getKey() {
        keyPressed = Keylistener.key;
    }

    /** This will change the position of the current block */
    private void switchOrientation() {
        // changes the orientation
        partCoords = keyPressed < 4 ? orientations[keyPressed] : partCoords;
        updateFarthestFrom();
    }

    private void updateFarthestFrom() {
        // every time the orentation is changed, the farthest part from the block is calculated on the x axis
        farthestFromBlockx = 0;
        for (int[] part: partCoords) {
            if (part[1] > farthestFromBlockx) {
                farthestFromBlockx = part[1];
            }
        }

        // every time the orentation is changed, the farthest part from the block is calculated on the y axis
        farthestFromBlocky = 0;
        for (int[] part: partCoords) {
            if (part[0] > farthestFromBlocky) {
                farthestFromBlocky = part[0];
            }
        }
    }

    /**
     * Shifts the block in the specified direction (keyboard input)
     *
     * @param boardWidth - the width of the board
     */
    private void shift(int boardWidth) {
        // shifts the block in the specified direction
        switch (keyPressed) {
            case 4:
                xpos--;
                break;
            case 5:
                xpos++;
                break;
            case 6:
                System.out.println("Slammin!");
                slam();
        }

        // constantly shoves the piece back into the board if it is out of bounds
        while (xpos + farthestFromBlockx >= boardWidth - 1) {
            xpos--;
        }

        // constantly shoves the piece back into the board if it is out of bounds
        while (xpos < 0 ) {
            xpos++;
        }
    }

    /** 
     * Returns the current coordinate position of the top left position of the block
     *
     * @return coords - the coordinates of the block
     */
    public int[] getCoords() {
        return new int[] {ypos, xpos};
    }

    /**
     * Returns the coordinates of the adjacent pieces that make up the block reletive to the block's position
     *
     * @return partCoords - the array of coordinates for the specified orientation of the block
     */
    public int[][] getAdjacentCoords() {
        return partCoords;
    }

    /** Causes the block to fall */
    public Boolean fall() {
        // variable checking if the block has touched the bottom
        boolean isNotTouchingBottom = (ypos + farthestFromBlocky < board.boardHeight - 1);
        // variable checking if the block will touch a block in the next gameloop
        boolean willTouchBlock = false;

        // loops through each coordinate in the current orientation
        for (int[] part: partCoords) {
            // checks if there is a block under each part of the block
             willTouchBlock = board.getSpace(
                xpos + part[1], 
                // checks below the part if it is not at the bottom layer (it would throw an error if it tried to check for a spot that did not exist)
                (ypos + part[0] == 24 ? 24: ypos + part[0] + 1));
        }

        // if the part is not touching the bottom and will not touch a block in the next gameloop, allow it to move down
        if (isNotTouchingBottom && !willTouchBlock) {
            ypos++;
        } else {
            return false;
        }
        return true;
    }

    public void slam() {
        while (fall()) {
            updateFarthestFrom();
        };
    }

    /** The main blockloop where everything is run */
    public void blockLoop(int boardHeight, int boardWidth) {
        getKey();
        switchOrientation();
        shift(boardWidth);
        // set the key to a value that is not being use, so the block does not continue to repeat the same action
        Keylistener.key = 7;
    }
}

File: TetPy/Blocks/block_abc.py
class Block_ABC:

    def __init__(self, board_width, board_height):
        self.xpos = 0
        self.ypos = 0
        self.part_coords = [[]]
        self.orientations = [[[]]]

        self.board_width = board_width
        self.board_height = board_height

        self.farthest_from_block_x = 0
        self.farthest_from_block_y = 0

    def get_coords(self):
        return [self.xpos, self.ypos]

    def get_adjacent_coords(self):
        return self.part_coords

    def take_action(self, action):
        # takes the action based on the input
        if action == "left":
            self.move_left()

        elif action == "right":
            self.move_right()

        elif action == "w":
            self.part_coords = self.orientations[0]

        elif action == "a":
            self.part_coords = self.orientations[1]

        elif action == "s":
            self.part_coords = self.orientations[2]

        elif action == "d":
            self.part_coords = self.orientations[3]

        # every time the orentation is changed, the farthest part from the block is calculated on the x axis
        self.farthest_from_block_x = 0;
        for part in self. part_coords:
            if part[0] > self.farthest_from_block_x:
                self.farthest_from_block_x = part[0]

        # every time the orentation is changed, the farthest part from the block is calculated on the y axis
        self.farthest_from_block_y = 0;
        for part in self.part_coords:
            if part[1] > self.farthest_from_block_y:
                self.farthest_from_block_y = part[1];

    def fall(self, board):
        # variable checking if the block has touched the bottom
        is_not_touching_bottom = (self.ypos + self.farthest_from_block_y < self.board_height - 1);
        # variable to check if the block will touch another block in the next loop
        #will_touch_block = False

        # loops through each coordinate in the current orientation
        for part in self.part_coords:

            y_space = 0
            # checks below the part if it is not at the bottom layer (it would throw an error if it tried to check for a spot that did not exist)
            if self.ypos + part[1] == 24:
                y_space = 24
            else:
                y_space = self.ypos + part[1] + 1


            # checks if there is a block under each part of the block
            will_touch_block = board.get_space(
                self.xpos + part[0],
                y_space
            )

        # if the part is not touching the bottom and will not touch a block in the next gameloop, allow it to move down
        if is_not_touching_bottom and not will_touch_block:
            self.ypos += 1
            return True;
        else:
            return False

    def move_left(self):
        self.xpos -= 1

        if self.xpos < 0:
            self.xpos += 1

    def move_right(self):
        self.xpos += 1

        for piece in self.part_coords:
            if (self.xpos + piece[0]) >= self.board_width - 1:
                self.xpos -= 1
